File: python/controllers/corpus.py
import json
import uuid
from datetime import datetime

from flask import Blueprint, request
from mongoengine import *

from models.corpus_sentence import Corpussentence
from models.parallel_corpus import Parallelcorpus
from models.response import CustomResponse
from models.sentence_log import Sentencelog
from models.single_corpus import Singlecorpus
from models.status import Status

corpus_api = Blueprint('corpus_api', __name__, url_prefix='/v1')
BASE_CORPUS = 'BASE_CORPUS'
STATUS_ACTIVE = 'ACTIVE'


@corpus_api.route("/upload-corpus", methods=['POST'])
def upload_corpus():
    global BASE_CORPUS, STATUS_ACTIVE
    basename = str(uuid.uuid4())
    f_corpus = request.files['corpus']
    name = request.form.getlist('name')
    domain = request.form.getlist('domain')
    lang = request.form.getlist('lang')
    if lang is None or len(lang) == 0 or name is None or len(name) == 0 or len(name[0]) == 0 or domain is None or len(
            domain) == 0 or len(domain[0]) == 0 or request.files is None or request.files['corpus'] is None:
        res = CustomResponse(
            Status.ERR_GLOBAL_MISSING_PARAMETERS.value, None)
        return res.getres(), Status.ERR_GLOBAL_MISSING_PARAMETERS.value['http']['status']
    description = ''
    if request.form.getlist('description') is not None and len(request.form.getlist('description')) > 0 and len(
            request.form.getlist('description')[0]) > 0:
        description = request.form.getlist('description')[0]
    corpus = Singlecorpus(status=STATUS_ACTIVE, created_on=datetime.now(), name=name[0], corpusid=basename,
                          domain=domain[0], lang=lang[0], description=description,
                          created_by=request.headers.get('ad-userid'))
    corpus.tags = [BASE_CORPUS, lang[0]]
    corpus.save()
    index = 0
    for f in f_corpus:
        sentence = Corpussentence(sentence=str(f), index=index)
        sentence.tags = [basename, lang[0]]
        sentence.original = True
        sentence.parallelcorpusid = []
        sentence.created_by = request.headers.get('ad-userid')
        sentence.save()
        index += 1
    res = CustomResponse(Status.SUCCESS.value, None)
    return res.getres()


@corpus_api.route("/get-corpus", methods=['GET'])
def get_corpus_list():
    global STATUS_ACTIVE
    corpuses = Singlecorpus.objects(status=STATUS_ACTIVE).to_json()
    res = CustomResponse(Status.SUCCESS.value, json.loads(corpuses))
    return res.getres()


@corpus_api.route("/get-parallel-corpus", methods=['GET'])
def get_parallel_corpus_list():
    global STATUS_ACTIVE
    corpuses = Parallelcorpus.objects(status=STATUS_ACTIVE).to_json()
    res = CustomResponse(Status.SUCCESS.value, json.loads(corpuses))
    return res.getres()


@corpus_api.route("/get-parallel-corpus-sentences", methods=['GET'])
def get_parallel_corpus_sentences_list():
    basename = request.args.get('basename')
    if basename is None or len(basename) == 0:
        res = CustomResponse(
            Status.ERR_GLOBAL_MISSING_PARAMETERS.value, None)
        return res.getres(), Status.ERR_GLOBAL_MISSING_PARAMETERS.value['http']['status']
    parallel_corpus = Parallelcorpus.objects(basename=basename)
    if parallel_corpus is None or len(parallel_corpus) == 0:
        res = CustomResponse(
            Status.DATA_NOT_FOUND.value, None)
        return res.getres(), Status.DATA_NOT_FOUND.value['http']['status']
    parallel_corpus_dict = json.loads(parallel_corpus.to_json())
    source_sentences = Corpussentence.objects.filter(
        Q(tags=parallel_corpus_dict[0]['source_id']) & Q(parallelcorpusid=basename)).order_by(
        'index')
    target_sentences = Corpussentence.objects.filter(
        Q(tags=parallel_corpus_dict[0]['target_id']) & Q(parallelcorpusid=basename)).order_by(
        'index')
    data = {'source': json.loads(source_sentences.to_json()), 'target': json.loads(target_sentences.to_json())}
    res = CustomResponse(Status.SUCCESS.value, data)
    return res.getres()


@corpus_api.route("/update-corpus-sentences", methods=['POST'])
def update_sentences():
    body = request.get_json()
    source_sentence = ''
    source_edited_sentence = ''
    target_sentence = ''
    target_edited_sentence = ''
    if body['id'] is None or (body['sentences'] is None or not isinstance(body['sentences'], list)):
        res = CustomResponse(
            Status.ERR_GLOBAL_MISSING_PARAMETERS.value, None)
        return res.getres(), Status.ERR_GLOBAL_MISSING_PARAMETERS.value['http']['status']
    for sentence in body['sentences']:
        if sentence['source'] is not None:
            source = sentence['source']
            if '_id' in source and source['_id']['$oid'] is not None:
                source_db = Corpussentence.objects(_id=source['_id']['$oid'])
                if source_db is not None:
                    source_db_dict = json.loads(source_db.to_json())
                    source_sentence = source_db_dict[0]['sentence']
                    source_edited_sentence = source['sentence']
                    if source_db_dict[0]['original']:
                        parallelcorpusid = source_db_dict[0]['parallelcorpusid']
                        if body['id'] in parallelcorpusid:
                            parallelcorpusid.remove(body['id'])
                            source_db.update(parallelcorpusid=parallelcorpusid)
                            sentence = Corpussentence(sentence=source['sentence'], index=source_db_dict[0]['index'])
                            sentence.tags = source_db_dict[0]['tags']
                            sentence.original = False
                            sentence.parallelcorpusid = [body['id']]
                            sentence.created_by = request.headers.get('ad-userid')
                            sentence.save()
                    else:
                        source_db.update(sentence=source['sentence'])
                target = sentence['target']
                parallel_corpus = Parallelcorpus.objects(basename=body['id'])
                parallel_corpus_dict = json.loads(parallel_corpus.to_json())
                target_sentences = Corpussentence.objects(
                    Q(tags=parallel_corpus_dict[0]['target_id']) & Q(index=source['index']))
                target_edited_sentence = target['sentence']
                if target_sentences is not None and len(target_sentences) > 0:
                    target_sentence = json.loads(target_sentences.to_json())[0]['sentence']
                    target_sentences.update(set__sentence=target['sentence'])
                else:
                    sentence = Corpussentence(sentence=target['sentence'], index=source['index'])
                    sentence.tags = [parallel_corpus_dict[0]['target_id'], parallel_corpus_dict[0]['target_lang']]
                    sentence.original = False
                    sentence.parallelcorpusid = [body['id']]
                    sentence.created_by = request.headers.get('ad-userid')
                    sentence.save()
            sentence_log = Sentencelog(source_words=source_sentence.split(" "), target_words=target_sentence.split(" "),
                                       source_edited_words=source_edited_sentence.split(" "),
                                       updated_on=datetime.now(), edited_by=request.headers.get('ad-userid'),
                                       parent_id=sentence['source']['_id']['$oid'],
                                       target_edited_words=target_edited_sentence.split(" "),
                                       basename=body['id'], source=source_sentence, target=target_sentence,
                                       source_edited=source_edited_sentence, target_edited=target_edited_sentence)
            sentence_log.save()
    res = CustomResponse(Status.SUCCESS.value, None)
    return res.getres()


@corpus_api.route("/create-parallel-corpus", methods=['POST'])
def create_parallel_corpus():
    global STATUS_ACTIVE
    body = request.get_json()
    if body['source_corpus'] is None or len(body['source_corpus']) == 0:
        res = CustomResponse(
            Status.ERR_GLOBAL_MISSING_PARAMETERS.value, None)
        return res.getres(), Status.ERR_GLOBAL_MISSING_PARAMETERS.value['http']['status']
    basename = str(uuid.uuid4())
    target_corpus_id = str(uuid.uuid4())
    source = body['source_corpus']
    name = body['name']
    domain = body['domain']
    target_lang = body['target_lang']
    source_lang = body['source_lang']
    corpus = Singlecorpus(status=STATUS_ACTIVE, created_on=datetime.now(), name=name, corpusid=target_corpus_id,
                          domain=domain, lang=target_lang, created_by=request.headers.get('ad-userid'))
    corpus.tags = [BASE_CORPUS, target_lang]
    corpus.save()
    source_corpus = Singlecorpus.objects(corpusid=source)
    if source_corpus is None or len(source_corpus) == 0:
        res = CustomResponse(
            Status.DATA_NOT_FOUND.value, None)
        return res.getres(), Status.DATA_NOT_FOUND.value['http']['status']
    parallel_corpus = Parallelcorpus(source_lang=source_lang, target_lang=target_lang, name=name, domain=domain,
                                     basename=basename, source_id=source, target_id=target_corpus_id,
                                     status=STATUS_ACTIVE)
    parallel_corpus.save()
    source_sentences = Corpussentence.objects(Q(tags=source) & Q(original=True))
    for source_sentence in source_sentences:
        source_sentence_dict = json.loads(source_sentence.to_json())
        source_sentence_tags = source_sentence_dict['parallelcorpusid']
        source_sentence_tags.append(basename)
        source_sentence.parallelcorpusid = source_sentence_tags
        source_sentence.save()
    res = CustomResponse(Status.SUCCESS.value, None)
    return res.getres()

File: node/controllers/corpus.js
/*
 * @Author: ghost 
 * @Date: 2019-08-29 17:32:50 
 * @Last Modified by: aroop.ghosh@tarento.com
 * @Last Modified time: 2019-08-30 14:22:51
 */
var Corpus = require('../models/corpus');
var SentenceLog = require('../models/sentencelog');
var Sentence = require('../models/sentence');
var Response = require('../models/response')
var APIStatus = require('../errors/apistatus')
var StatusCode = require('../errors/statuscodes').StatusCode
var Status = require('../utils/status').Status
var KafkaProducer = require('../kafka/producer');
var LOG = require('../logger/logger').logger
var UUIDV4 = require('uuid/v4')
var async = require('async');
var fs = require("fs");
var ParallelCorpus = require('../models/parallelCorpus');
var Sentence = require('../models/sentence')
const CJSON = require('circular-json');
const { Translate } = require('@google-cloud/translate');
const projectId = "anuvaad";

var COMPONENT = "corpus";
const STATUS_ACCEPTED = "ACCEPTED"
const STATUS_COMPLETED = "COMPLETED"
const STATUS_REJECTED = "REJECTED"
const PATH_HUMAN_CORRECTION = 'human_correction'
const TOPIC_STAGE_3 = 'searchreplace'
var PARALLEL_CORPUS_COMPONENT = "parallelCorpus";

const LANGUAGES = {
    'Hindi': 'hi',
    'English': 'en',
    'Bengali': 'bn',
    'Gujarati': 'gu',
    'Marathi': 'mr',
    'Kannada': 'kn',
    'Telugu': 'te',
    'Malayalam': 'ml',
    'Punjabi': 'pa',
    'Tamil': 'ta',
    'Urdu': 'ur',
}


exports.fetchCorpus = function (req, res) {
    Corpus.fetchAll((err, corpus) => {
        if (err) {
            let apistatus = new APIStatus(StatusCode.ERR_GLOBAL_SYSTEM, COMPONENT).getRspStatus()
            return res.status(apistatus.http.status).json(apistatus);
        }
        let response = new Response(StatusCode.SUCCESS, corpus).getRsp()
        return res.status(response.http.status).json(response);
    })
}

exports.fetchCorpusSentences = function (req, res) {
    var basename = req.query.basename
    var pagesize = req.query.pagesize
    var pageno = req.query.pageno
    var status = req.query.status
    if (basename == null || basename.length == 0) {
        let apistatus = new APIStatus(StatusCode.ERR_GLOBAL_MISSING_PARAMETERS, PARALLEL_CORPUS_COMPONENT).getRspStatus()
        return res.status(apistatus.http.status).json(apistatus);
    }
    Corpus.findOne({ basename: basename }, function (error, corpus) {
        Sentence.countDocuments({ basename: basename }, function (err, count) {
            Sentence.fetch(basename, pagesize, pageno, status, null, function (err, sentences) {
                if (err) {
                    let apistatus = new APIStatus(StatusCode.ERR_GLOBAL_SYSTEM, COMPONENT).getRspStatus()
                    return res.status(apistatus.http.status).json(apistatus);
                }

                if (sentences && Array.isArray(sentences) && sentences.length > 0) {
                    let sentences_arr = []
                    let target_lang = 'en'
                    target_lang = LANGUAGES[corpus['_doc']['target_lang']] ? LANGUAGES[corpus['_doc']['target_lang']] : 'en'
                    sentences.map((sentence) => {
                        sentences_arr.push(sentence._doc.source)
                    })
                    return translateFromGoogle(target_lang, sentences_arr, sentences, res, count)
                }
                let response = new Response(StatusCode.SUCCESS, sentences).getRsp()
                return res.status(response.http.status).json(response);
            })
        })
    })
}

var translateFromGoogle = function (targetlang, text, sentences, res, count) {
    const translate = new Translate({
        projectId: projectId,
    });

    // Translates some text into English
    translate
        .translate(text, targetlang)
        .then(results => {
            let sentencearr = []
            results[0].map((r, index) => {
                var s = sentences[index]
                s['_doc']['translation'] = r
                sentencearr.push(s)
            })
            let response = new Response(StatusCode.SUCCESS, sentencearr, count).getRsp()
            return res.status(response.http.status).json(response);
        })
        .catch(err => {
            LOG.error(err)
            let response = new Response(StatusCode.SUCCESS, sentences).getRsp()
            return res.status(response.http.status).json(response);
        });
}

exports.updateSentences = function (req, res) {
    //Check required params
    if (!req.body || !req.body.sentences || !Array.isArray(req.body.sentences)) {
        let apistatus = new APIStatus(StatusCode.ERR_GLOBAL_MISSING_PARAMETERS, COMPONENT).getRspStatus()
        return res.status(apistatus.http.status).json(apistatus);
    }
    async.each(req.body.sentences, function (sentence, callback) {
        LOG.debug("Updating sentence [%s]", JSON.stringify(sentence))
        Sentence.find({ _id: sentence._id }, {}, function (err, results) {
            if (results && Array.isArray(results) && results.length > 0) {
                var sentencedb = results[0]
                let userId = req.headers['ad-userid']
                let sentencelog = { edited_by: userId, status: sentencedb._doc.status, source_words: sentencedb._doc.source.split(' '), target_words: sentencedb._doc.target.split(' '), source_edited_words: sentence.source.split(' '), target_edited_words: sentence.target.split(' '), updated_on: new Date(), parent_id: sentencedb._doc._id, basename: sentencedb._doc.basename, source: sentencedb._doc.source, source_edited: sentence.source, target: sentencedb._doc.target, target_edited: sentence.target }
                SentenceLog.save([sentencelog], (err, results) => {
                    if (err) {
                        LOG.error(err)
                        callback()
                    }
                    Sentence.updateSentence(sentence, (error, results) => {
                        if (error) {
                            LOG.error(error)
                            callback()
                        }
                        LOG.debug("Sentence updated [%s]", JSON.stringify(sentence))
                        callback()
                    })
                })

            } else {
                callback('data not found')
            }
        })
    }, function (err) {
        if (err) {
            let apistatus = new APIStatus(StatusCode.ERR_GLOBAL_SYSTEM, COMPONENT).getRspStatus()
            return res.status(apistatus.http.status).json(apistatus);
        }
        let apistatus = new APIStatus(StatusCode.SUCCESS, COMPONENT).getRspStatus()
        return res.status(apistatus.http.status).json(apistatus);
    });
}

exports.updateSentencesStatus = function (req, res) {
    //Check required params
    if (!req.body || !req.body.sentences || !Array.isArray(req.body.sentences)) {
        let apistatus = new APIStatus(StatusCode.ERR_GLOBAL_MISSING_PARAMETERS, COMPONENT).getRspStatus()
        return res.status(apistatus.http.status).json(apistatus);
    }
    let corpus_basename = req.body.sentences[0].basename
    async.each(req.body.sentences, function (sentence, callback) {
        LOG.debug("Updating sentence status [%s]", JSON.stringify(sentence))
        Sentence.find({ _id: sentence._id }, {}, function (err, results) {
            if (results && Array.isArray(results) && results.length > 0) {
                var sentencedb = results[0]
                let userId = req.headers['ad-userid']
                let sentencelog = { edited_by: userId, source: sentencedb._doc.source, target: sentencedb._doc.target, is_status_changed: true, updated_on: new Date(), parent_id: sentencedb._doc._id, basename: sentencedb._doc.basename, status: sentencedb._doc.status, status_edited: sentence.status }
                SentenceLog.save([sentencelog], (err, results) => {
                    if (err) {
                        LOG.error(err)
                        callback()
                    }
                    Sentence.updateSentenceStatus(sentence, (error, results) => {
                        if (error) {
                            LOG.error(error)
                            callback()
                        }
                        LOG.debug("Sentence updated [%s]", JSON.stringify(sentence))
                        callback()
                    })
                })

            }
            else {
                LOG.debug('Data not found')
                callback('data not found')
            }
        })
    }, function (err) {
        if (err) {
            let apistatus = new APIStatus(StatusCode.ERR_GLOBAL_SYSTEM, COMPONENT).getRspStatus()
            return res.status(apistatus.http.status).json(apistatus);
        }
        Corpus.findByCondition({ basename: corpus_basename }, function (err, corpus) {
            if (err || !corpus) {
                let apistatus = new APIStatus(StatusCode.ERR_GLOBAL_SYSTEM, COMPONENT).getRspStatus()
                return res.status(apistatus.http.status).json(apistatus);
            }
            let corpus_obj = corpus[0]._doc
            Sentence.countDocuments({ basename: corpus_basename }, function (err, totalcount) {
                Sentence.countDocuments({ basename: corpus_basename, $or: [{ status: STATUS_ACCEPTED }, { status: STATUS_REJECTED }] }, function (err, processedcount) {
                    if (processedcount == totalcount) {
                        LOG.info("Data processed")
                        corpus_obj.data_processed = true
                        corpus_obj.status = STATUS_COMPLETED
                        KafkaProducer.getInstance().getProducer((err, producer) => {
                            if (err) {
                                LOG.error("Unable to connect to KafkaProducer");
                            } else {
                                LOG.debug("KafkaProducer connected")
                                let payloads = [
                                    {
                                        topic: TOPIC_STAGE_3, messages: JSON.stringify({ data: { session_id: corpus_obj.basename }, path: PATH_HUMAN_CORRECTION }), partition: 0
                                    }
                                ]
                                producer.send(payloads, function (err, data) {
                                    LOG.debug('Produced')
                                });
                            }
                        })
                        Corpus.updateCorpus(corpus_obj, function (err, doc) {
                            if (err) {
                                LOG.error(err)
                            } else {
                                LOG.info('Corpus updated')
                            }
                        })
                    }
                    let apistatus = new APIStatus(StatusCode.SUCCESS, COMPONENT).getRspStatus()
                    return res.status(apistatus.http.status).json(apistatus);
                })
            })
        })
    });
}

exports.translateSource = function (req, res) {
    var basename = req.query.basename
    var source = req.query.source
    if (basename == null || basename.length == 0 || !source || source.length == 0) {
        let apistatus = new APIStatus(StatusCode.ERR_GLOBAL_MISSING_PARAMETERS, PARALLEL_CORPUS_COMPONENT).getRspStatus()
        return res.status(apistatus.http.status).json(apistatus);
    }
    Corpus.findOne({ basename: basename }, function (error, corpus) {
        if (corpus) {
            let target_lang = 'en'
            target_lang = corpus['_doc'] && LANGUAGES[corpus['_doc']['target_lang']] ? LANGUAGES[corpus['_doc']['target_lang']] : 'en'
            const translate = new Translate({
                projectId: projectId,
            });
            translate
                .translate(source, target_lang)
                .then(results => {
                    let response = new Response(StatusCode.SUCCESS, [results[0]]).getRsp()
                    return res.status(response.http.status).json(response);
                })
                .catch(err => {
                    LOG.error(err)
                    let response = new Response(StatusCode.SUCCESS, sentences).getRsp()
                    return res.status(response.http.status).json(response);
                });
        }
        else {
            let response = new APIStatus(StatusCode.ERR_GLOBAL_NOTFOUND, COMPONENT).getRspStatus()
            return res.status(response.http.status).json(response);
        }
    })
}

exports.updateSentencesGrade = function (req, res) {
    //Check required params
    if (!req.body || !req.body.sentences || !Array.isArray(req.body.sentences)) {
        let apistatus = new APIStatus(StatusCode.ERR_GLOBAL_MISSING_PARAMETERS, COMPONENT).getRspStatus()
        return res.status(apistatus.http.status).json(apistatus);
    }
    async.each(req.body.sentences, function (sentence, callback) {
        LOG.debug("Updating sentence grade [%s]", JSON.stringify(sentence))
        if (sentence.rating || sentence.spelling_rating || sentence.context_rating || sentence.name_accuracy_rating || sentence.comments) {
            Sentence.find({ _id: sentence._id }, {}, function (err, results) {
                if (results && Array.isArray(results) && results.length > 0) {
                    var sentencedb = results[0]
                    let userId = req.headers['ad-userid']
                    let sentencelog = { edited_by: userId, source: sentencedb._doc.source, target: sentencedb._doc.target, is_grade_changed: true, updated_on: new Date(), parent_id: sentencedb._doc._id, basename: sentencedb._doc.basename, status: sentencedb._doc.status, name_accuracy_rating_edited: sentence.name_accuracy_rating, name_accuracy_rating: sentencedb._doc.name_accuracy_rating, spelling_rating_edited: sentence.spelling_rating, grade_edited: sentence.rating, grade: sentencedb._doc.rating, spelling_rating: sentencedb._doc.spelling_rating, context_rating: sentencedb._doc.context_rating, context_rating_edited: sentence.context_rating, comments: sentencedb._doc.comments, comments_edited: sentence.comments }
                    if (req.body.modelid) {
                        sentencelog.modelid = req.body.modelid
                    }
                    else if (sentencedb._doc.basename && sentencedb._doc.basename.indexOf('_') > 0) {
                        sentencelog.modelid = sentencedb._doc.basename.split('_')[1]
                    }
                    SentenceLog.save([sentencelog], (err, results) => {
                        if (err) {
                            LOG.error(err)
                            callback()
                        }
                        Sentence.updateSentenceGrade(sentence, (error, results) => {
                            callback()
                        })
                    })
                } else {
                    LOG.debug('Data not found')
                    callback('data not found')
                }
            })
        } else {
            LOG.debug('Rating not specified')
            callback()
        }
    }, function (err) {
        if (err) {
            let apistatus = new APIStatus(StatusCode.ERR_GLOBAL_SYSTEM, COMPONENT).getRspStatus()
            return res.status(apistatus.http.status).json(apistatus);
        }
        let apistatus = new APIStatus(StatusCode.SUCCESS, COMPONENT).getRspStatus()
        return res.status(apistatus.http.status).json(apistatus);
    });
}

exports.uploadCorpus = function (req, res) {
    //Check required params
    if (!req.files || req.files.length == 0 || !req.body || !req.body.name || !req.body.lang) {
        let apistatus = new APIStatus(StatusCode.ERR_GLOBAL_MISSING_PARAMETERS, COMPONENT).getRspStatus()
        return res.status(apistatus.http.status).json(apistatus);
    }
    let corpusid = UUIDV4()
    //Create corpus obj and save
    let corpus = { status: 'ACTIVE', created_on: new Date(), corpusid: corpusid, name: req.body.name, lang: req.body.lang }
    corpus.tags = ['BASE_CORPUS', req.body.lang]
    Corpus.saveCorpus(corpus, (err, doc) => {
        if (err) {
            let apistatus = new APIStatus(StatusCode.ERR_GLOBAL_SYSTEM, COMPONENT).getRspStatus()
            return res.status(apistatus.http.status).json(apistatus);
        }
        fs.readFile(req.files[0].path, 'utf8', function (err, data) {
            if (data.length > 0) {
                let sentences = data.split('\n')
                let sentencearr = []
                sentences.map((sentence, index) => {
                    let sentenceobj = { sentence: sentence, index: index }
                    sentenceobj.tags = [corpusid, req.body.lang]
                    sentenceobj.original = true
                    sentenceobj.parallelcorpusid = []
                    sentencearr.push(sentenceobj)
                })
                Sentence.saveSentences(sentencearr, (err, docs) => {
                    if (err) {
                        let apistatus = new APIStatus(StatusCode.ERR_GLOBAL_SYSTEM, COMPONENT).getRspStatus()
                        return res.status(apistatus.http.status).json(apistatus);
                    }
                    let apistatus = new APIStatus(StatusCode.SUCCESS, COMPONENT).getRspStatus()
                    return res.status(apistatus.http.status).json(apistatus);
                })
            }
        })
    })
}

exports.fetchParallelCorpus = function (req, res) {
    Corpus.fetchAll((err, corpus) => {
        if (err) {
            let apistatus = new APIStatus(StatusCode.ERR_GLOBAL_SYSTEM, COMPONENT).getRspStatus()
            return res.status(apistatus.http.status).json(apistatus);
        }
        let response = new Response(StatusCode.SUCCESS, corpus).getRsp()
        return res.status(response.http.status).json(response);
    })
}

exports.getParallelCorpusSentence = function (req, res) {
    var basename = req.query.basename
    if (basename == null || basename.length == 0) {
        let apistatus = new APIStatus(StatusCode.ERR_GLOBAL_MISSING_PARAMETERS, PARALLEL_CORPUS_COMPONENT).getRspStatus()
        return res.status(apistatus.http.status).json(apistatus);
    }
    ParallelCorpus.findById(basename, (err, result) => {
        if (err) {
            let apistatus = new APIStatus(StatusCode.ERR_GLOBAL_SYSTEM, COMPONENT).getRspStatus()
            return res.status(apistatus.http.status).json(apistatus);
        }

        sourceId = result.source_id
        targetId = result.target_id
        console.log("sourceId = " + sourceId)
        console.log("targetId = " + targetId)
        Sentence.findWithtagId(basename, sourceId, (err, source_sentences) => {
            if (err) {
                console.log("error while finding source sentence")
                let apistatus = new APIStatus(StatusCode.ERR_GLOBAL_SYSTEM, COMPONENT).getRspStatus()
                return res.status(apistatus.http.status).json(apistatus);
            }
            console.log("source sentences are " + CJSON.stringify(source_sentences))
            Sentence.findWithtagId(basename, targetId, (err, target_sentences) => {
                if (err) {
                    console.log("error while finding target sentence")
                    let apistatus = new APIStatus(StatusCode.ERR_GLOBAL_SYSTEM, COMPONENT).getRspStatus()
                    return res.status(apistatus.http.status).json(apistatus);
                }
                console.log("target sentences are " + target_sentences.sentence)
                let results = { 'source': source_sentences.sentence, 'target': target_sentences.sentence }
                let response = new Response(StatusCode.SUCCESS, results).getRsp()
                return res.status(response.http.status).json(response);
            })
        })
    })
}



